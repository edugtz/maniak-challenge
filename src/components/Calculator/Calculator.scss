.calculator-container {
    .range-slider-wrapper {
        margin-top: 22px;
        width: 100%;
    }

    .calculator-top-section-wrapper {
        display: flex;
        justify-content: space-between;
        align-items: flex-end;
        flex-wrap: wrap;

        .spending-description {
            p {
                margin: 0;
                font-size: $font-size;
                color: $black-color;
            }
        }

        .input-icon-wrapper {
            position: relative;
        
            .spending-icon {
                margin: 13px;
                display: flex;
                align-items: center;
                position: absolute;
                font-size: 24px;
                font-weight: 500;
                height: 26px;
                letter-spacing: 0.5px;
                color: rgba(22, 22, 22, 0.24);
                width: 15px;
            }
            
            .spending-input {
                border-radius: 4px;
                border: 1px solid #d6dcff;
                background-color: #fff;
                height: 52px;
                width: 190px;
                font-size: 36px;
                line-height: 1.11;
                letter-spacing: 0.8px;
                text-align: right;
                color: $black-color;
                font-weight: 500;
            
            }
        }
    }
    
    .calculator-mid-section-wrapper {
        display: flex;
        justify-content: space-between;
        align-items: flex-end;
        flex-wrap: wrap;
        margin-top: 66px;

        .employee-description {
            p {
                margin: 0;
                font-size: $font-size;
                color: $black-color;
            }
        }

        .input-wrapper {
            
            .employee-count-input {
                border-radius: 4px;
                border: 1px solid #d6dcff;
                background-color: #fff;
                height: 52px;
                width: 67px;
                font-size: 36px;
                line-height: 1.11;
                letter-spacing: 0.8px;
                text-align: right;
                color: $black-color;
                font-weight: 500;
            
            }
        }
    }

    .calculator-bottom-section-wrapper {
        margin-top: 86px;
        display: flex;
        justify-content: space-between
    }
}



