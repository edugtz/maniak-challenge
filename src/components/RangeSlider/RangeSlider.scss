.rc-slider {
    .rc-slider-track {
        background-color: $primary-color;
    }
    .rc-slider-rail {
        background-color: #f0f2ff;
    }
    .rc-slider-handle {
        border: 5px solid $primary-color;
        width: 16px;
        height: 16px;
        margin-top: -6px;

        &:focus {
            box-shadow: none;
            border: 5px solid $primary-color;
        }

        &:hover {
            border: 5px solid $primary-color;
        }
    }   
}