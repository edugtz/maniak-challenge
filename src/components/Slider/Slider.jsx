import React, { Component } from 'react';
import Arrows from './components/Arrows'
import Review from './components/Review'
import Indicator from './components/Indicator';
import './Slider.scss';

class Slider extends Component {
    constructor(props) {
        super(props);

        this.handleGoToPreviousReview = this.handleGoToPreviousReview.bind(this);
        this.handleGoToNextReview = this.handleGoToNextReview.bind(this);

        this.state = {
            activeIndex: 0
        };
    }

    handleGoToPreviousReview = (e) => {
        e.preventDefault();

        const { reviews } = this.props;
        let index = this.state.activeIndex;
        let reviewsLength = reviews.length;

        if(index < 1) {
            index = reviewsLength;
        }

        index--;

        this.setState(() => ({
            activeIndex: index
        }));
    }

    handleGoToNextReview = (e) => {
        e.preventDefault();

        const { reviews } = this.props;
        let index = this.state.activeIndex;
        let reviewsLength = reviews.length - 1;

        if (index === reviewsLength) {
            index = -1;
        }

        index++;

        this.setState(() => ({
            activeIndex: index
        }));
    }

    render() {
        const { reviews } = this.props;
        return(
            <div className="slider-container">
                <div className="review-container">
                    {reviews.length > 0 && reviews.map((review, index) => 
                        <Review 
                            key={`${review.name} - ${index}`}
                            index={index}
                            activeIndex={this.state.activeIndex}
                            review={review}
                        />
                    )}
                </div>
                <div className="controls-wrapper">
                        <Indicator
                            activeIndex={this.state.activeIndex + 1} 
                            total={reviews.length}
                        />
                        <Arrows 
                            handleGoToPreviousReview={this.handleGoToPreviousReview}
                            handleGoToNextReview={this.handleGoToNextReview}
                        />
                </div>
            </div>
        );
    }
} 

export default Slider;