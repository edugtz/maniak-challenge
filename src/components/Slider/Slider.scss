.controls-wrapper {
    display: flex;
    justify-content: flex-end;
    width: 100%;
    position: absolute;
    bottom: -10%;
}