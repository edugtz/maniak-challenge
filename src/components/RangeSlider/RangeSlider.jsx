import React from 'react';
import Slider from 'rc-slider';
import './RangeSlider.scss';
import 'rc-slider/assets/index.css';

const RangeSlider = props => {
    const { min, max, handleOnChange, defaultValue, value} = props;
    return (
        <Slider
            defaultValue={Number(defaultValue)} 
            min={min}
            max={max}
            onChange={handleOnChange}
            value={Number(value)}
        />
    );
};

export default RangeSlider;