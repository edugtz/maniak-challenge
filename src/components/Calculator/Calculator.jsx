import React, { Component } from 'react';
import './Calculator.scss';
import { faDollarSign } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import RangeSlider from '../RangeSlider';
import Savings from '../Savings';


class Calculator extends Component {
    constructor(props) {
        super(props);

        this.handleEmployeeChangeSlider = this.handleEmployeeChangeSlider.bind(this);
        this.handleSpendingChangeSlider = this.handleSpendingChangeSlider.bind(this);
        
        this.handleEmployeeChangeInput = this.handleEmployeeChangeInput.bind(this);
        this.handleSpendingChangeInput = this.handleSpendingChangeInput.bind(this);

        this.state = {
            employeeCount: 1,
            monthlySpending: 10,
            estimatedFoodCostSavings: 0,
            estimatedAnnualSavings: 0
        };
    };

    componentDidMount = () => {
        this.calculateSavings();
    };

    handleEmployeeChangeInput = (e) => {
        const value = e.target.value;

        if (value >= 1 && value <= 10) {
            this.setState({
                employeeCount: value
            }, () => {
                this.calculateSavings();
            });
        }
    };

    handleSpendingChangeInput = (e) => {
        const value = e.target.value;

        if (value >= 10 && value <= 100) {
            this.setState({
                monthlySpending: value
            }, () => {
                this.calculateSavings();
            });
        }
    };

    handleEmployeeChangeSlider = (value) => {
        this.setState(() => ({
            employeeCount: value
        }), () => {
            this.calculateSavings();
        });
    };

    handleSpendingChangeSlider = (value) => {
        this.setState(() => ({
            monthlySpending: value
        }), () => {
            this.calculateSavings();
        });
    };

    calculateSavings = () => {
        const { monthlySpending, employeeCount } = this.state;

        const estimatedFoodCostSavings = Number((monthlySpending) * 0.3).toFixed(1);
        const estimatedAnnualSavings = Number((employeeCount) * 1337 + estimatedFoodCostSavings).toFixed(1);

        this.setState(() => ({
            estimatedAnnualSavings,
            estimatedFoodCostSavings
        }));
    };

    render() {
        const { 
            monthlySpending, 
            employeeCount, 
            estimatedAnnualSavings, 
            estimatedFoodCostSavings 
        } = this.state;

        return(
            <div className="calculator-container">
                <div className="calculator-top-section-wrapper">
                    <div className="spending-description">
                        <p>Monthly <br></br> ingredient spending</p>
                    </div>

                    <div className="input-icon-wrapper">
                        <span className="spending-icon"><FontAwesomeIcon icon={faDollarSign} /></span>
                        <input className="spending-input" 
                            value={monthlySpending} 
                            onChange={this.handleSpendingChangeInput} 
                            type="text"    
                        />
                    </div>
                    <div className="range-slider-wrapper">
                        <RangeSlider
                            id="spending-slider"
                            handleOnChange={this.handleSpendingChangeSlider} 
                            value={monthlySpending}
                            defaultValue={10}
                            min={10}
                            max={100}
                        />
                    </div>
                </div>
                <div className="calculator-mid-section-wrapper">
                    <div className="employee-description">
                        <p>Full time employees that <br></br> process invoices</p>
                    </div>

                    <div className="input-wrapper">
                        <input className="employee-count-input" 
                            value={employeeCount} 
                            onChange={this.handleEmployeeChangeInput} 
                            type="text"    
                        />
                    </div>
                    <div className="range-slider-wrapper">
                        <RangeSlider
                            id="employee-slider" 
                            handleOnChange={this.handleEmployeeChangeSlider}
                            value={employeeCount}
                            defaultValue={1}
                            min={1}
                            max={10}
                        />
                    </div>
                </div>
                <div className="calculator-bottom-section-wrapper">
                    <Savings 
                        amount={estimatedFoodCostSavings} 
                        description="Estimated food cost savings"
                    />
                    <Savings 
                        amount={estimatedAnnualSavings} 
                        description="Your estimated annual savings"    
                    />
                </div>
            </div>
        );
    }
};

export default Calculator;