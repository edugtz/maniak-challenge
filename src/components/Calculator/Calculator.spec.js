import React from 'react';
import Calculator from './Calculator';
import RangeSlider from '../../components/RangeSlider';
import Slider from 'rc-slider';
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import { createSerializer } from "enzyme-to-json";
import sinon from 'sinon';

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));
configure({ adapter: new Adapter() });

const sandbox = sinon.createSandbox();

describe('Calculator component', () => {

    afterEach(() => {
        sandbox.restore();
    });

    const buildWrapper = (props) => {
        return shallow(<Calculator {...props} />);
    };

    // Snapshot test
    it('should render correctly', () => {
        const wrapper = buildWrapper({ });

        expect(wrapper).toMatchSnapshot();
    });

    it('should update and display employee count correctly', () => {
        const employeeCount = 8;
        const wrapper = buildWrapper({ });
        const employeeSliderSpy = sandbox.spy(wrapper.instance(), 'handleEmployeeChangeSlider');
        
        wrapper.instance().forceUpdate();

        wrapper.find(RangeSlider)
            .at(1)
            .dive()
            .find(Slider)
            .simulate('change', employeeCount);

        
        expect(employeeSliderSpy.calledOnce).toBe(true);
        expect(wrapper.state().employeeCount).toEqual(employeeCount);
    });

    it('should update and display monthly spendings correctly', () => {
        const monthlySpending = 90;
        const wrapper = buildWrapper({ });
        const monthlySpendingSpy = sandbox.spy(wrapper.instance(), 'handleSpendingChangeSlider');
        
        wrapper.instance().forceUpdate();

        wrapper.find(RangeSlider)
            .at(0)
            .dive()
            .find(Slider)
            .simulate('change', 90);

        
        expect(monthlySpendingSpy.calledOnce).toBe(true);
        expect(wrapper.state().monthlySpending).toEqual(monthlySpending);
    });
});
