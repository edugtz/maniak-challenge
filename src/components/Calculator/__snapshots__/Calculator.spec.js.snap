// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component should render correctly 1`] = `
<div
  className="calculator-container"
>
  <div
    className="calculator-top-section-wrapper"
  >
    <div
      className="spending-description"
    >
      <p>
        Monthly 
        <br />
         ingredient spending
      </p>
    </div>
    <div
      className="input-icon-wrapper"
    >
      <span
        className="spending-icon"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                288,
                512,
                Array [],
                "f155",
                "M209.2 233.4l-108-31.6C88.7 198.2 80 186.5 80 173.5c0-16.3 13.2-29.5 29.5-29.5h66.3c12.2 0 24.2 3.7 34.2 10.5 6.1 4.1 14.3 3.1 19.5-2l34.8-34c7.1-6.9 6.1-18.4-1.8-24.5C238 74.8 207.4 64.1 176 64V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48h-2.5C45.8 64-5.4 118.7.5 183.6c4.2 46.1 39.4 83.6 83.8 96.6l102.5 30c12.5 3.7 21.2 15.3 21.2 28.3 0 16.3-13.2 29.5-29.5 29.5h-66.3C100 368 88 364.3 78 357.5c-6.1-4.1-14.3-3.1-19.5 2l-34.8 34c-7.1 6.9-6.1 18.4 1.8 24.5 24.5 19.2 55.1 29.9 86.5 30v48c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-48.2c46.6-.9 90.3-28.6 105.7-72.7 21.5-61.6-14.6-124.8-72.5-141.7z",
              ],
              "iconName": "dollar-sign",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          title=""
          transform={null}
        />
      </span>
      <input
        className="spending-input"
        onChange={[Function]}
        type="text"
        value={10}
      />
    </div>
    <div
      className="range-slider-wrapper"
    >
      <RangeSlider
        defaultValue={10}
        handleOnChange={[Function]}
        id="spending-slider"
        max={100}
        min={10}
        value={10}
      />
    </div>
  </div>
  <div
    className="calculator-mid-section-wrapper"
  >
    <div
      className="employee-description"
    >
      <p>
        Full time employees that 
        <br />
         process invoices
      </p>
    </div>
    <div
      className="input-wrapper"
    >
      <input
        className="employee-count-input"
        onChange={[Function]}
        type="text"
        value={1}
      />
    </div>
    <div
      className="range-slider-wrapper"
    >
      <RangeSlider
        defaultValue={1}
        handleOnChange={[Function]}
        id="employee-slider"
        max={10}
        min={1}
        value={1}
      />
    </div>
  </div>
  <div
    className="calculator-bottom-section-wrapper"
  >
    <Savings
      amount="3.0"
      description="Estimated food cost savings"
    />
    <Savings
      amount="13373.0"
      description="Your estimated annual savings"
    />
  </div>
</div>
`;
